File: src/main/kotlin/yandex_algo_training/contest01/diego03/Solution.java
package yandex_algo_training.contest01.diego03;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.TreeSet;

public class Solution {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int diegoNum = Integer.parseInt(br.readLine());
        String diegoStr = br.readLine().trim();
        TreeSet<Integer> diegoStickers = new TreeSet<>();
        if (diegoNum > 0) {
            String[] nums = diegoStr.split(" ");
            for (String num : nums) {
                diegoStickers.add(Integer.parseInt(num));
            }
        }
        int numCollections = Integer.parseInt(br.readLine());
        String collStr = br.readLine();
        List<Integer> colls = new ArrayList<>();
        if (numCollections > 0) {
            String[] nums = collStr.split(" ");
            for (String num : nums) {
                colls.add(Integer.parseInt(num));
            }
        }
        List<Integer> res = countStickers(colls, diegoStickers);
        res.forEach(System.out::println);
        br.close();
    }

    private static List<Integer> countStickers(List<Integer> colls, TreeSet<Integer> diego) {
        if (colls.isEmpty()) return Collections.emptyList();
        List<Integer> res = new ArrayList<>();
        if (diego.isEmpty()) {
            for (int i = 0; i < colls.size(); ++i) {
                res.add(0);
            }
            return res;
        }
        for (Integer num : colls) {
            res.add(diego.headSet(num).size());
        }
        return res;
    }
}

File: src/main/kotlin/yandex_algo_training/contest01/diego03/solution.kt
package yandex_algo_training.contest01.diego03

fun main() {
    val diegoNum = readLine()!!.toInt()
    val diegoStr = readLine()!!.trim()
    val diegoStickers = if (diegoNum == 0) {
        emptyList<Int>()
    } else {
        diegoStr.split(" ").toSet().map { it.toInt() }.sorted()
    }
    val numCollectionaires = readLine()!!.toInt()
    val collectionairesStr = readLine()!!
    val collectionaireStickers = if (numCollectionaires == 0) {
        emptyList<Int>()
    } else {
        collectionairesStr.split(" ").map { it.toInt() }
    }
    val answer = countNeededStickers(collectionaireStickers, diegoStickers)
    answer.forEach { println(it) }
}

fun countNeededStickers(
    collectionaireStickers: List<Int>,
    diegoStickers: List<Int>
): List<Int> {
    if (collectionaireStickers.isEmpty()) return emptyList()
    if (diegoStickers.isEmpty()) {
        return List(collectionaireStickers.size) { 0 }
    }
    val result = mutableListOf<Int>()
    collectionaireStickers.forEach { s ->
        if (s > diegoStickers[diegoStickers.lastIndex]) {
            result.add(diegoStickers.size)
        } else {
            val index = leftBinSearch<Int>(0, diegoStickers.size - 1) { idx ->
                diegoStickers[idx] >= s
            }

            result.add(index)
        }

    }
    return result
}

fun <T> leftBinSearch(
    left: Int,
    right: Int,
    check: (Int) -> Boolean
): Int {
    var l = left
    var r = right
    while (l < r) {
        val m = l + (r - l) / 2
        if (check(m)) { // acceptable
            r = m
        } else { // not acceptable
            l = m + 1
        }
    }
    return l
}

